### PR TITLE
8317980: Optimization for Integer.parseInt and Long.parseLong 

### PR DESCRIPTION
**Pull request by wenshao**
Currently, about 25% of the time spent by the Integer.parseInt and Long.parseLong methods is spent on the Character.digit method.

The Character.digit method is common to all radixes. We can use a digit method optimized for Latin1 encoding radix 10 to improve the performance of Integer.parseInt and Long.parseLong methods.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8317980](https://bugs.openjdk.org/browse/JDK-8317980): Optimization for Integer.parseInt and Long.parseLong (**Enhancement** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20168/head:pull/20168` \
`$ git checkout pull/20168`

Update a local copy of the PR: \
`$ git checkout pull/20168` \
`$ git pull https://git.openjdk.org/jdk.git pull/20168/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20168`

View PR using the GUI difftool: \
`$ git pr show -t 20168`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20168.diff">https://git.openjdk.org/jdk/pull/20168.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20168#issuecomment-2226547618)

### CODE REVIEW

**Comment by wenshao** [comment]
The performance numbers are as follows

## MacBook M1 Pro
```diff
# baseline 
-Benchmark          (size)  Mode  Cnt  Score   Error  Units
-Longs.parseLong       500  avgt   15  2.613 ? 0.020  us/op
-Integers.parseInt     500  avgt   15  2.303 ? 0.006  us/op

+Benchmark          (size)  Mode  Cnt  Score   Error  Units
+Longs.parseLong       500  avgt   15  2.473 ? 0.020  us/op +5.66%
+Integers.parseInt     500  avgt   15  2.219 ? 0.002  us/op +3.78%
```

## MacBook 2018 i9
```diff
-Benchmark          (size)  Mode  Cnt  Score   Error  Units
-Integers.parseInt     500  avgt   15  3.114 ? 0.058  us/op
-Longs.parseLong       500  avgt   15  3.569 ? 0.059  us/op

+Benchmark          (size)  Mode  Cnt  Score   Error  Units
+Integers.parseInt     500  avgt   15  3.038 ? 0.033  us/op  +2.50%
+Longs.parseLong       500  avgt   15  3.065 ? 0.032  us/op +16.44%
```

**Comment by liach** [comment]
A straightforward guess would be converting the `Character.digit()` call to a utility method like:
```java
private static int digit(char ch/*, int radix*/) {
    if (ch >= '0' && ch <= '9') // or replace it with something like Math.min('9', '0' + radix)
        return ch - '0'; // ascii base-10 fast path
    return digit(ch/*, radix*/);
}
```
Inspired by this: https://github.com/openjdk/jdk/blob/ae9f318fc35eeab497e546ebab9faed6ec774ec5/src/java.base/share/classes/jdk/internal/constant/ConstantUtils.java#L236

However I don't know how JIT has compiled the code, so not sure how my speculative approach works compared to yours. But it is definitely somewhat less invasive.

**Comment by wenshao** [comment]
> A straightforward guess would be converting the `Character.digit()` call to a utility method like:
> 
> ```java
> private static int digit(char ch/*, int radix*/) {
>     if (ch >= '0' && ch <= '9') // or replace it with something like Math.min('9', '0' + radix)
>         return ch - '0'; // ascii base-10 fast path
>     return digit(ch/*, radix*/);
> }
> ```
> 
> Inspired by this:
> 
> https://github.com/openjdk/jdk/blob/ae9f318fc35eeab497e546ebab9faed6ec774ec5/src/java.base/share/classes/jdk/internal/constant/ConstantUtils.java#L236
> 
> However I don't know how JIT has compiled the code, so not sure how my speculative approach works compared to yours. But it is definitely somewhat less invasive, without the new table and latin1 requirements.

In the parse scenario, I made many attempts and found no method faster than table lookup, because the input length will be greater than 1 and the table will be looked up multiple times. In this scenario, I think it is correct to use table lookup.

File: src/java.base/share/classes/java/lang/CharacterDataLatin1.java.template [hunk 1]
```diff
@@ -251,6 +252,26 @@ class CharacterDataLatin1 extends CharacterData {
                 && radix <= Character.MAX_RADIX) ? value : -1;
     }
 
+    @Stable
```
Comment (jddarcy) [inline src/java.base/share/classes/java/lang/CharacterDataLatin1.java.template:255]: Please add a textual description of the semantics of this array.

**Comment by wenshao** [comment]
@jddarcy  do I need to make more improvements to this PR? Thanks

**Pull request closed without merging**